consensus/src/simplex/actors/batcher/actor.rs

```diff
@@ -594,35 +494,63 @@ impl<
             };
             timer.observe();
 
-            // Send messages to voter
+            // Process verified votes
             let batch = voters.len() + failed.len();
-            trace!(%view, batch, "batch verified messages");
+            trace!(%view, batch, "batch verified votes");
             self.verified.inc_by(batch as u64);
             self.batch_size.observe(batch as f64);
-            for voter in voters {
-                consensus.verified(voter).await;
-            }
 
             // Block invalid signers
-            if !failed.is_empty() {
-                for invalid in failed {
-                    if let Some(signer) = self.participants.key(invalid) {
-                        warn!(?signer, "blocking peer");
-                        self.blocker.block(signer.clone()).await;
-                    }
+            for invalid in failed {
+                if let Some(signer) = self.participants.key(invalid) {
+                    warn!(?signer, "blocking peer for invalid signature");
+                    self.blocker.block(signer.clone()).await;
                 }
             }
 
+            // Store verified votes for certificate construction
+            let round = work.get_mut(&view).expect("round must exist");
+            for valid in voters {
+                round.add_verified(valid);
+            }
+
+            // Try to construct and forward certificates
+            let mut recover_timer = self.recover_latency.timer();
+            if let Some(notarization) = round.try_construct_notarization(&self.scheme) {
+                recover_timer.observe();
+                debug!(%view, "constructed notarization, forwarding to voter");
+                voter
+                    .recovered(Certificate::Notarization(notarization))
+                    .await;
+            } else {
+                recover_timer.cancel();
+            }
+            let mut recover_timer = self.recover_latency.timer();
+            if let Some(nullification) = round.try_construct_nullification(&self.scheme) {
+                recover_timer.observe();
+                debug!(%view, "constructed nullification, forwarding to voter");
+                voter
+                    .recovered(Certificate::Nullification(nullification))
+                    .await;
+            } else {
+                recover_timer.cancel();
+            }
+            let mut recover_timer = self.recover_latency.timer();
```

Comment:
nit: this `recover_timer` 3x (that also requires an `else` statement attached) almost seems a bit of a code-smell. Wondering it it's possible to put it in the `try_construct_*` functions themselves

Reply:
hmmm interesting idea